File: Medium/0002.AddTwoNumber/addtwonumber.ml
type listnode = { v : int; next : listnode option }

let add_two_numbers (l1 : listnode option) (l2 : listnode option) : listnode option =
  (* Recursive helper: consumes nodes and carry, returns the result list. *)
  let rec loop a b carry =
    match (a, b, carry) with
    | (None, None, 0) ->
        None  (* nothing left *)
    | _ ->
        let av, anext =
          match a with
          | None -> (0, None)
          | Some n -> (n.v, n.next)
        in
        let bv, bnext =
          match b with
          | None -> (0, None)
          | Some n -> (n.v, n.next)
        in
        let sum = av + bv + carry in
        let digit = sum mod 10 in
        let carry' = sum / 10 in
        Some { v = digit; next = loop anext bnext carry' }
  in
  loop l1 l2 0

File: Medium/0002-add-two-numbers/addtwonumber.scala
// LeetCode-style definition:
class ListNode(var x: Int = 0) {
  var next: ListNode = null
}

object Solution {
  def addTwoNumbers(l1: ListNode, l2: ListNode): ListNode = {

    // Recursive helper: builds the result list from current nodes + carry.
    def loop(a: ListNode, b: ListNode, carry: Int): ListNode = {
      // Base case: nothing left to add
      if (a == null && b == null && carry == 0) null
      else {
        val av = if (a != null) a.x else 0
        val bv = if (b != null) b.x else 0

        val sum = av + bv + carry
        val node = new ListNode(sum % 10)        // Current digit
        node.next = loop(
          if (a != null) a.next else null,
          if (b != null) b.next else null,
          sum / 10                                // Next carry
        )
        node
      }
    }

    loop(l1, l2, 0)
  }
}
